src/main/scala/org/broadinstitute/hail/Utils.scala

```diff
@@ -572,6 +592,68 @@ class RichPairTraversableOnce[K, V](val t: TraversableOnce[(K, V)]) extends AnyV
   }
 }
 
+class RichPairIterator[K, V](val it: Iterator[(K, V)]) extends AnyVal {
+
+  def localKeySort[T](implicit ord: Ordering[T], kOrd: Ordering[K], ev: (K) => T): Iterator[(K, V)] = {
+    lazyKSorted[T](it)
+  }
+
+  /**
+    * Precondition: the iterator is T-sorted. Moreover, ev must be monotonic. We lazily K-sort each block of
+    * T-equivalent elements.
+    */
+  private def lazyKSorted[T](it: Iterator[(K, V)])
+    (implicit tOrd: Ordering[T], kOrd: Ordering[K], ev: (K) => T): Iterator[(K, V)] =
+  if (it.isEmpty)
+    Iterator.empty
+  else {
+    val nextValue = it.next
+    val t = ev(nextValue._1)
+    val (tEquivalents, greater) = it.span({ case (k, _) => tOrd.equiv(t, ev(k)) })
+    val kSorted = (Iterator(nextValue) ++ tEquivalents).toSeq.sortBy(_._1)
+    // NB: ++ is lazy in its second argument
+    kSorted.iterator ++ lazyKSorted[T](greater)
```

Comment:
This does waaay too much consing.  I think it should be a single iterator with a single priority queue.  I will suggest an implementation...